tmt/utils.py

```diff
@@ -3607,6 +3604,62 @@ def __init__(
 NormalizeCallback = Callable[[Any, tmt.log.Logger], T]
 
 
+def dataclass_normalize_field(
+        container: Any,
+        keyname: str,
+        raw_value: Any,
+        logger: tmt.log.Logger) -> Any:
+    """
+    Normalize and assign a value to container field.
+
+    If there is a normalization callback defined for the field, either in field
+    metadata or as a special ``_normalize_$keyname`` method, the method is
+    called for ``raw_value``, and its return value is assigned to container
+    field instead of ``value``.
+    """
+
+    normalize_callback: Optional[NormalizeCallback[Any]] = None
+
+    # First try new-style fields, i.e. normalize callback stored in field metadata
+    if dataclasses.is_dataclass(container):
+        field: dataclasses.Field[Any] = dataclass_field_by_name(container, keyname)
+        normalize_callback = dataclass_field_metadata(field).normalize_callback
+
+    if not normalize_callback:
+        normalize_callback = getattr(container, f'_normalize_{keyname}', None)
+
+    if normalize_callback:
+        value = normalize_callback(raw_value, logger)
+
+    else:
+        value = raw_value
+
+    # Set attribute by adding it to __dict__ directly. Messing with setattr()
+    # might cause re-use of mutable values by other instances.
+    container.__dict__[keyname] = value
+
+    # TODO: keep this code for debugging purpose - when logging changes with
+    # https://github.com/teemtee/tmt/pull/1685, this would be turned into
```

Comment:
Logging changes in #1685 have been already merged. Shall we update this part to use proper logging calls as suggested? Or keep it as is to enable only when really needed?

Reply:
Changed in cce47a92ffa216c736d25acd9066d8e6ba52e5ea - logging is enabled, but with a deep debug level.

Reply:
Nice, thanks!